### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by brettmadrid**
Starting the Spring Challenge

### CODE REVIEW

File: client/src/components/Jokes.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  componentDidMount() {
+    axios
+      .get("/jokes")
+      .then(res => {
+        this.setState({ jokes: res.data });
+      })
+      .catch(e => {
+        console.error();
+      });
```
Comment (johnoro) [inline client/src/components/Jokes.js:18]: I assume the error `e` is meant to passed to the inner function here, but why use the extra step? Just use the inner function as your callback:
```javascript
  .catch(console.error);
```

File: client/src/components/Login.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { Button, Form, FormGroup, Label, Input, Spinner } from "reactstrap";
+
+class Login extends React.Component {
+  state = {
+    username: "",
+    password: "",
+  };
+
+  handleInputChange = event => {
+    const { name, value } = event.target;
```
Comment (johnoro) [inline client/src/components/Login.js:12]: There are a few alternative ways you could use destructuring here to the same effect:
- taking target off of the event parameter (so as to slightly reduce repetition)
```javascript
handleInputChange = ({ target }) => {
  const { name, value } = target;
  this.setState({
    [name]: value
  })
};
```
- or, take off name and value since you don't need target (note that you _won't_ have access to the target object here, but only the innermost properties you destructure off, in this case name and value):
```javascript
handleInputChange = ({ target: { name, value } }) => {
  this.setState({
    [name]: value
  })
};
```
nested destructuring can get a little hard to follow, so you can also format it differently to make it shorter in terms of line length and to read:
```javascript
handleInputChange = ({
  target: { name, value }
}) => {
  this.setState({
    [name]: value
  })
};
```

File: client/src/auth/requiresAuth.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+import axios from "axios";
+
+axios.defaults.baseURL = "http://localhost:9000/api";
```
Comment (johnoro) [inline client/src/auth/requiresAuth.js:4]: You could export an axios object with this so that you don't have to repeat yourself elsewhere:
```javascript
import axios from "axios";

axios.defaults.baseURL = "http://localhost:9000/api";

module.exports = axios;
```
or, using an [axios instance](https://github.com/axios/axios#creating-an-instance):
```javascript
import axios from "axios";

const axiosInstance = axios.create({
  baseURL: "http://localhost:9000/api"
});

module.exports = axiosInstance;
```

File: client/src/components/Registration.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+              name="username"
+              id="username"
+              placeholder="username"
+              value={this.state.username}
+              onChange={this.handleInputChange}
+              type="text"
+            />
+          </FormGroup>
+          <FormGroup className="mb-2 mr-sm-2 mb-sm-0">
+            <Label for="password" className="mr-sm-2">
```
Comment (johnoro) [inline client/src/components/Registration.js:49]: I _think_ `htmlFor` is typically used here since `for` is a JavaScript keyword.

File: config/users-model.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const [id] = await db('users').insert(user);
+
+  return findById(id);
+}
+
+function findById(id) {
+  return db('users')
+    .select('id', 'username')
+    .where({ id })
+    .first();
```
Comment (johnoro) [inline config/users-model.js:23]: The design of this function with excluding passwords might make you think that you can't use your `findBy` helper function, but you _can_ call `select` after other queries:
```javascript
  return findBy({ id })
    .select('id', 'username')
    .first();
```

**Review by johnoro — APPROVED** [review]
Looking awesome!

**Comment by brettmadrid** [comment]
This makes total sense.  The code was already there and I just assumed it worked efficiently.

File: client/src/components/Registration.js (hunk 4 above)
Comment (brettmadrid) [reply client/src/components/Registration.js:49]: Oh wow, good catch!  looks like a typo.  I wonder why it still works?  strange!

**Comment by brettmadrid** [comment]
The axios object makes a lot of sense.

File: client/src/components/Jokes.js (hunk 1 above)
Comment (brettmadrid) [reply client/src/components/Jokes.js:18]: simpler

File: client/src/components/Login.js (hunk 2 above)
Comment (brettmadrid) [reply client/src/components/Login.js:12]: Wow, these are all very cool! Thanks for showing me these alternate ways!
